File: CPP/No354.cc
/**
 * Created by Xiaozhong on 2020/8/29.
 * Copyright (c) 2020/8/29 Xiaozhong. All rights reserved.
 */
#include "vector"
#include "algorithm"

using namespace std;

class Solution {
public:
    int maxEnvelopes(vector<vector<int>> &envelopes) {
        // 把数据按照 w 升序， h 降序排列
        sort(envelopes.begin(), envelopes.end(), [](vector<int> &a, vector<int> &b) {
            return a[0] == b[0] ? a[1] > b[1] : a[0] < b[0];
        });

        vector<int> ans;
        for (const auto &e:envelopes) {
            // 二分查找第一个大于等于的地方
            auto p = lower_bound(ans.begin(), ans.end(), e[1]);
            if (p == ans.end()) ans.push_back(e[1]);
            else *p = e[1];
        }
        return ans.size();
    }
};

int main() {
    Solution s;
}

File: Java/src/No354.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

public class No354 {
    /*public int maxEnvelopes(int[][] envelopes) {
        int size = envelopes.length;
        if (size == 0) return 0;
        Arrays.sort(envelopes, new Comparator<int[]>() {
            @Override
            public int compare(int[] o1, int[] o2) {
                return o1[0] == o2[0] ? o2[1] - o1[1] : o1[0] - o2[0];
            }
        });

        List<Integer> ans = new ArrayList<>();
        ans.add(envelopes[0][1]);
        for (int i = 1; i < size; ++i) {
            int num = envelopes[i][1];
            if (num > ans.get(ans.size() - 1)) {
                ans.add(num);
            } else {
                int idx = binarySearch(ans, num);
                ans.set(idx, num);
            }
        }
        return ans.size();
    }

    private int binarySearch(List<Integer> list, int num) {
        int left = 0, right = list.size() - 1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (list.get(mid) < num) left = mid + 1;
            else right = mid;
        }
        return left;
    }*/
    public int maxEnvelopes(int[][] envelopes) {
        int n = envelopes.length;
        int[] dp = new int[n + 1];
        Arrays.sort(envelopes, (o1, o2) -> {
            return o2[0] - o1[0] == 0 ? o2[1] - o1[1] : o2[0] - o1[0];
        });
        int ans = 0;
        for (int i = 0; i < n; ++i) {
            dp[i] = 1;
            for (int j = 0; j < i; ++j) {
                if (envelopes[j][0] > envelopes[i][0] && envelopes[j][1] > envelopes[i][1]) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
            ans = Math.max(dp[i], ans);
        }
        return ans;
    }

    public static void main(String[] args) {
        int[][] envelopes = {{5, 4}, {6, 4}, {6, 7}, {2, 3}};
        No354 no354 = new No354();
        System.out.println(no354.maxEnvelopes(envelopes));
    }
}
